### PR TITLE
controller: deflake controller factory test

### PR DESCRIPTION
**Pull request by mfojtik**
Rework the shutdown test to use channels instead time.Sleep, make it more reliable and flake prone.

### CODE REVIEW

File: pkg/controller/factory/controller_test.go [hunk 1]
```diff
@@ -106,51 +106,52 @@ func TestResyncController(t *testing.T) {
...
+		case 1:
 			syncContext.Queue().Add("TestKey1")
 			syncContext.Queue().Add("TestKey2")
 			syncContext.Queue().Add("TestKey3")
 			syncContext.Queue().Add("TestKey4")
-			syncContext.Queue().Add("TestKey5")
+		case 5:
+			close(allWorkersBusy)
 		}
 		syncCallCountMutex.Unlock()
```
Comment (sttts) [inline pkg/controller/factory/controller_test.go]: move up behind `++`?

File: pkg/controller/factory/controller_test.go [hunk 2]
```diff
@@ -106,51 +106,52 @@ func TestResyncController(t *testing.T) {
...
 
-	shutdown() // initiate controller shutdown
-	<-runComplete
+	// this blocks until all workers are shut down.
+	controller.Run(controllerCtx, 5)
 
 	workersShutdownMutex.Lock()
-	if workersShutdownCount != 6 {
-		t.Fatalf("expected all 6 workers to gracefully shutdown, got %d", workersShutdownCount)
+	if workersShutdownCount != 5 {
```
Comment (sttts) [inline pkg/controller/factory/controller_test.go]: how do you know that all 5 have finished the sync func?

File: pkg/controller/factory/controller_test.go (hunk 1 above)
Comment (mfojtik) [reply pkg/controller/factory/controller_test.go]: you need syncCallCount in switch

File: pkg/controller/factory/controller_test.go (hunk 2 above)
Comment (mfojtik) [reply pkg/controller/factory/controller_test.go]: you count them when they finish :)

Comment (sttts) [reply pkg/controller/factory/controller_test.go]: There is a race.

Comment (sttts) [reply pkg/controller/factory/controller_test.go]: Didn't realize `controller.Run(controllerCtx, 5)` blocks.

**Pull request merged**